File: services/marketplace-service/src/main/java/com/designer/marketplace/lms/entity/Quiz.java
package com.designer.marketplace.lms.entity;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.data.mongodb.core.index.Indexed;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
@Document(collection = "quizzes")
public class Quiz {

    @Id
    private String id;
    
    @Indexed
    private String courseId;
    
    @Indexed
    private String lessonId;
    
    private String title;
    private String description;
    
    @Builder.Default
    private List<Question> questions = new ArrayList<>();
    
    private Integer passingScore;  // Percentage required to pass (0-100)
    private Integer timeLimit;     // Time limit in minutes (null = no limit)
    private Integer maxAttempts;   // Max attempts allowed (null = unlimited)
    
    @Builder.Default
    private Boolean shuffleQuestions = false;
    
    @Builder.Default
    private Boolean shuffleOptions = false;
    
    @Builder.Default
    private Boolean showCorrectAnswers = true;  // After submission
    
    @Builder.Default
    private Boolean isActive = true;
    
    private LocalDateTime createdAt;
    private LocalDateTime updatedAt;
    
    public void addQuestion(Question question) {
        if (question.getId() == null) {
            question.setId(UUID.randomUUID().toString());
        }
        questions.add(question);
    }
    
    public int getTotalPoints() {
        return questions.stream()
                .mapToInt(q -> q.getPoints() != null ? q.getPoints() : 1)
                .sum();
    }
    
    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class Question {
        private String id;
        private QuestionType type;
        private String questionText;
        private String explanation;  // Shown after answering
        
        @Builder.Default
        private List<Option> options = new ArrayList<>();
        
        @Builder.Default
        private List<String> correctAnswers = new ArrayList<>();  // For multiple choice/multi-select
        
        private String correctTextAnswer;  // For fill-in-blank
        
        @Builder.Default
        private Integer points = 1;
        
        private Integer orderIndex;
    }
    
    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class Option {
        private String id;
        private String text;
        private Boolean isCorrect;
        private Integer orderIndex;
    }
    
    public enum QuestionType {
        MULTIPLE_CHOICE,      // Single correct answer
        MULTI_SELECT,         // Multiple correct answers
        TRUE_FALSE,           // True/False
        FILL_IN_BLANK,        // Text answer
        MATCHING,             // Match pairs
        ORDERING              // Put in correct order
    }
}

File: services/lms-service/Models/Quiz.cs
using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;

namespace LmsService.Models;

public class Quiz
{
    [BsonId]
    [BsonRepresentation(BsonType.ObjectId)]
    public string Id { get; set; } = null!;

    [BsonElement("courseId")]
    public string CourseId { get; set; } = null!;

    [BsonElement("lessonId")]
    public string LessonId { get; set; } = null!;

    [BsonElement("title")]
    public string Title { get; set; } = null!;

    [BsonElement("description")]
    public string? Description { get; set; }

    [BsonElement("questions")]
    public List<QuizQuestion> Questions { get; set; } = new();

    [BsonElement("passingScore")]
    public double PassingScore { get; set; } = 70.0;

    [BsonElement("timeLimitMinutes")]
    public int? TimeLimitMinutes { get; set; }

    [BsonElement("maxAttempts")]
    public int MaxAttempts { get; set; } = 3;

    [BsonElement("shuffleQuestions")]
    public bool ShuffleQuestions { get; set; }

    [BsonElement("shuffleOptions")]
    public bool ShuffleOptions { get; set; }

    [BsonElement("showCorrectAnswers")]
    public bool ShowCorrectAnswers { get; set; } = true;

    [BsonElement("createdAt")]
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

    [BsonElement("updatedAt")]
    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;
}

public class QuizQuestion
{
    [BsonElement("id")]
    public string Id { get; set; } = Guid.NewGuid().ToString();

    [BsonElement("text")]
    public string Text { get; set; } = null!;

    [BsonElement("type")]
    public QuestionType Type { get; set; }

    [BsonElement("options")]
    public List<QuizOption> Options { get; set; } = new();

    [BsonElement("correctAnswer")]
    public string? CorrectAnswer { get; set; }

    [BsonElement("explanation")]
    public string? Explanation { get; set; }

    [BsonElement("points")]
    public int Points { get; set; } = 1;

    [BsonElement("orderIndex")]
    public int OrderIndex { get; set; }
}

public class QuizOption
{
    [BsonElement("id")]
    public string Id { get; set; } = Guid.NewGuid().ToString();

    [BsonElement("text")]
    public string Text { get; set; } = null!;

    [BsonElement("isCorrect")]
    public bool IsCorrect { get; set; }
}

public enum QuestionType
{
    MultipleChoice,
    TrueFalse,
    ShortAnswer,
    MultiSelect
}

public class QuizAttempt
{
    [BsonId]
    [BsonRepresentation(BsonType.ObjectId)]
    public string Id { get; set; } = null!;

    [BsonElement("quizId")]
    public string QuizId { get; set; } = null!;

    [BsonElement("studentId")]
    public long StudentId { get; set; }

    [BsonElement("enrollmentId")]
    public string EnrollmentId { get; set; } = null!;

    [BsonElement("answers")]
    public List<QuizAnswer> Answers { get; set; } = new();

    [BsonElement("score")]
    public double Score { get; set; }

    [BsonElement("totalPoints")]
    public int TotalPoints { get; set; }

    [BsonElement("earnedPoints")]
    public int EarnedPoints { get; set; }

    [BsonElement("passed")]
    public bool Passed { get; set; }

    [BsonElement("startedAt")]
    public DateTime StartedAt { get; set; }

    [BsonElement("completedAt")]
    public DateTime? CompletedAt { get; set; }

    [BsonElement("timeTakenSeconds")]
    public int? TimeTakenSeconds { get; set; }
}

public class QuizAnswer
{
    [BsonElement("questionId")]
    public string QuestionId { get; set; } = null!;

    [BsonElement("selectedOptions")]
    public List<string> SelectedOptions { get; set; } = new();

    [BsonElement("textAnswer")]
    public string? TextAnswer { get; set; }

    [BsonElement("isCorrect")]
    public bool IsCorrect { get; set; }

    [BsonElement("pointsEarned")]
    public int PointsEarned { get; set; }
}
